File: backend/api/schemas/strategy_runs.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel


class StrategyRunRequest(BaseModel):
    strategy_id: str
    provider: str
    symbol: str
    timeframe: str
    start: datetime
    end: datetime
    asset_class: str = "equity"
    exchange: str = "NASDAQ"
    parameters: dict[str, Any] = {}


class SignalResponse(BaseModel):
    strategy_id: str
    timestamp: datetime
    symbol: str
    timeframe: str
    signal_type: str
    entry_reference: float
    invalidation_level: float
    confidence: float | None = None
    reasoning: str | None = None


class ForecastResponse(BaseModel):
    generated_at: datetime
    target_timestamp: datetime
    target_price: float
    direction: str
    confidence: float
    invalidation_price: float | None = None
    reason: str | None = None


class IndicatorPointResponse(BaseModel):
    timestamp: datetime
    value: float


class IndicatorSeriesResponse(BaseModel):
    name: str
    kind: str
    pane: str
    color: str | None = None
    points: list[IndicatorPointResponse]


class StrategyRunResponse(BaseModel):
    strategy_id: str
    status: str
    candles_received: int
    signals: list[SignalResponse]
    forecasts: list[ForecastResponse]
    indicators: list[IndicatorSeriesResponse] = []
    warnings: list[str]
    error: str | None = None

File: frontend/src/api/strategyRuns.ts
export interface StrategyRunRequest {
  strategy_id: string
  provider: string
  symbol: string
  timeframe: string
  start: string
  end: string
  asset_class?: string
  exchange?: string
  parameters?: Record<string, unknown>
}

export interface SignalResponse {
  strategy_id: string
  timestamp: string
  symbol: string
  timeframe: string
  signal_type: 'long' | 'short' | 'exit' | 'reduce'
  entry_reference: number
  invalidation_level: number
  confidence: number | null
  reasoning: string | null
}

export interface ForecastResponse {
  generated_at: string
  target_timestamp: string
  target_price: number
  direction: 'long' | 'short' | 'neutral'
  confidence: number
  invalidation_price: number | null
  reason: string | null
}

export interface IndicatorPoint {
  timestamp: string
  value: number
}

export interface IndicatorSeries {
  name: string
  kind: string
  pane: string
  color: string | null
  points: IndicatorPoint[]
}

export interface StrategyRunResponse {
  strategy_id: string
  status: 'success' | 'empty' | 'failed'
  candles_received: number
  signals: SignalResponse[]
  forecasts: ForecastResponse[]
  indicators: IndicatorSeries[]
  warnings: string[]
  error: string | null
}

import { authedFetch } from './client'

export async function runStrategy(req: StrategyRunRequest): Promise<StrategyRunResponse> {
  const resp = await authedFetch('/strategy-runs/run', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(req),
  })
  if (!resp.ok) {
    const err = await resp.json().catch(() => ({ detail: resp.statusText }))
    throw new Error(err.detail ?? `HTTP ${resp.status}`)
  }
  return resp.json()
}
